src/components/views/rooms/DecryptionFailureBar.tsx

```diff
@@ -0,0 +1,247 @@
+        defaultDispatcher.dispatch(payload);
+    };
+
+    const onResetClick = (): void => {
+        const store = SetupEncryptionStore.sharedInstance();
+        store.resetConfirm()
+            .then(() => context.downloadKeys([context.getUserId()], true))
+            .then(() => updateDeviceInfo());
+    };
+
+    const statusIndicator = waiting ? <Spinner /> : <div className="mx_DecryptionFailureBar_icon" />;
+
+    let headline: JSX.Element;
+    let body: JSX.Element;
+    let button: JSX.Element;
+    if (waiting) {
+        headline = <React.Fragment>
+            { _t("Decrypting messages...") }
+        </React.Fragment>;
+        body = <React.Fragment>
+            { _t("Please wait as we try to decrypt your messages. This may take a few moments.") }
+        </React.Fragment>;
+    } else if (needsVerification) {
+        if (hasOtherVerifiedDevices || hasKeyBackup) {
+            headline = <React.Fragment>
+                { _t("Verify this device to access all messages") }
+            </React.Fragment>;
+            body = <React.Fragment>
+                { _t("This device was unable to decrypt some messages because it has not been verified yet.") }
+            </React.Fragment>;
+            button = <AccessibleButton kind="primary" onClick={onVerifyClick}>
+                { _t("Verify") }
+            </AccessibleButton>;
+        } else {
+            headline = <React.Fragment>
+                { _t("Reset your keys to prevent future decryption errors") }
+            </React.Fragment>;
+            body = <React.Fragment>
+                { _t(
+                    "You will not be able to access old undecryptable messages, " +
+                    "but resetting your keys will allow you to receive new messages.",
+                ) }
+            </React.Fragment>;
+            button = <AccessibleButton kind="primary" onClick={onResetClick}>
+                { _t("Reset") }
+            </AccessibleButton>;
+        }
+    } else if (hasOtherVerifiedDevices) {
+        headline = <React.Fragment>
+            { _t("Open another device to load encrypted messages") }
+        </React.Fragment>;
+        body = <React.Fragment>
+            { _t(
+                "This device is requesting decryption keys from your other devices. " +
+                "Opening one of your other devices may speed this up.",
+            ) }
+        </React.Fragment>;
+        button = <AccessibleButton kind="primary_outline" onClick={onDeviceListClick}>
+            { _t("View your device list") }
+        </AccessibleButton>;
+    } else {
+        headline = <React.Fragment>
+            { _t("Requesting keys to decrypt messages...") }
+        </React.Fragment>;
+        body = <React.Fragment>
+            { _t("Some messages could not be decrypted. This device is requesting keys from the messages' senders.") }
+        </React.Fragment>;
```

Comment:
given that, post-RHUL, we never reshare keys with other users, I think this is inaccurate.

Reply:
That's not true. Post-RHUL, we no longer accept *unsolicited* keys shared by other users (unless it's part of being invited to a new room). However, we do still send key requests, and we accept keys that are shared in response to a specific request we know we sent.

Reply:
Ah, looks like there's been another change when I wasn't watching, and now post-post-RHUL we don't request keys from other users at all :/ I'll update the text.